File: packages/typescript/service/tests/disclosure.test.ts
import { describe, it, expect } from "vitest";
import { resolveDisclosureLevel } from "../src/disclosure.js";

describe("Fixed mode", () => {
  it("full returns full", () => {
    expect(resolveDisclosureLevel("full", {})).toBe("full");
  });
  it("reduced returns reduced", () => {
    expect(resolveDisclosureLevel("reduced", {})).toBe("reduced");
  });
  it("redacted returns redacted", () => {
    expect(resolveDisclosureLevel("redacted", {})).toBe("redacted");
  });
  it("ignores token claims in fixed mode", () => {
    expect(
      resolveDisclosureLevel("redacted", { "anip:caller_class": "internal" }, { internal: "full" }),
    ).toBe("redacted");
  });
});

describe("Policy mode", () => {
  it("resolves from caller class", () => {
    expect(
      resolveDisclosureLevel("policy", { "anip:caller_class": "internal" }, { internal: "full", default: "redacted" }),
    ).toBe("full");
  });
  it("falls back to default", () => {
    expect(
      resolveDisclosureLevel("policy", { "anip:caller_class": "unknown" }, { internal: "full", default: "reduced" }),
    ).toBe("reduced");
  });
  it("falls back to redacted when no default", () => {
    expect(
      resolveDisclosureLevel("policy", { "anip:caller_class": "unknown" }, { internal: "full" }),
    ).toBe("redacted");
  });
  it("no token claim uses default", () => {
    expect(
      resolveDisclosureLevel("policy", {}, { internal: "full", default: "reduced" }),
    ).toBe("reduced");
  });
  it("no policy returns redacted", () => {
    expect(
      resolveDisclosureLevel("policy", { "anip:caller_class": "internal" }),
    ).toBe("redacted");
  });
  it("null claims uses default", () => {
    expect(
      resolveDisclosureLevel("policy", null, { default: "reduced" }),
    ).toBe("reduced");
  });
});

File: packages/go/service/disclosure_test.go
package service

import "testing"

func TestResolveDisclosureLevel_FixedModes(t *testing.T) {
	for _, level := range []string{"full", "reduced", "redacted"} {
		got := ResolveDisclosureLevel(level, nil, nil)
		if got != level {
			t.Errorf("ResolveDisclosureLevel(%q) = %q, want %q", level, got, level)
		}
	}
}

func TestResolveDisclosureLevel_PolicyNoPolicy(t *testing.T) {
	got := ResolveDisclosureLevel("policy", nil, nil)
	if got != "redacted" {
		t.Errorf("policy mode with nil policy should default to redacted, got %q", got)
	}
}

func TestResolveDisclosureLevel_PolicyWithCallerClass(t *testing.T) {
	claims := map[string]any{"anip:caller_class": "internal"}
	policy := map[string]string{
		"internal": "full",
		"default":  "redacted",
	}
	got := ResolveDisclosureLevel("policy", claims, policy)
	if got != "full" {
		t.Errorf("internal caller should get full, got %q", got)
	}
}

func TestResolveDisclosureLevel_PolicyFallsBackToDefault(t *testing.T) {
	claims := map[string]any{"anip:caller_class": "unknown_class"}
	policy := map[string]string{
		"internal": "full",
		"default":  "reduced",
	}
	got := ResolveDisclosureLevel("policy", claims, policy)
	if got != "reduced" {
		t.Errorf("unknown caller should fall back to default, got %q", got)
	}
}

func TestResolveDisclosureLevel_PolicyFromScope(t *testing.T) {
	claims := map[string]any{
		"scope": []any{"travel.search", "audit:full"},
	}
	policy := map[string]string{
		"audit_full": "full",
		"default":    "redacted",
	}
	got := ResolveDisclosureLevel("policy", claims, policy)
	if got != "full" {
		t.Errorf("audit:full scope should resolve to audit_full class, got %q", got)
	}
}
